guides/inotify-watch-limits.md

```diff
@@ -0,0 +1,96 @@
+---
+title: Inotify Watch Limit Increase
+description: Learn how to increase the inotify watcher limit.
+---
+
+When running a webpack project, you may encounter an error similar to the
+following:
+
+```text
+Watchpack Error (watcher): Error: ENOSPC: System limit for number of file
+watchers reached, watch '/some/path'
+```
+
+This results from a low number of [inotify
+watches](https://confluence.jetbrains.com/display/IDEADEV/Inotify+Watches+Limit)
+combined with high node usage, causing the log stream to fail.
+
+## Resolution
+
+Increase the number of file watchers. Because the setting quantifying the number
+of file watchers isn't namespaced, you'll need to raise the maximum number at
+the node level.
+
+One way to do this is to use a daemonset in the cluster with a privileged
+container to set the maximum number of file watchers:
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: more-fs-watchers
+  namespace: kube-system
+  labels:
+    app: more-fs-watchers
+    k8s-app: more-fs-watchers
+spec:
+  selector:
+    matchLabels:
+      k8s-app: more-fs-watchers
+  template:
+    metadata:
+      labels:
+        name: more-fs-watchers
+        k8s-app: more-fs-watchers
+    spec:
+      hostNetwork: true
+      hostPID: true
+      hostIPC: true
+      initContainers:
+        - command:
+            - sh
```

Comment:
@scsmithr How come this uses `sh -c` instead of directly running the `sysctl` command?

Reply:
I tested this and it works - there's only so much we can do since we need a privileged container to access /proc/sys, but I at least tried to cut down on the host namespaces we were sharing into the DaemonSet pods:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: more-fs-watchers
  namespace: kube-system
  labels:
    app: more-fs-watchers
    k8s-app: more-fs-watchers
spec:
  selector:
    matchLabels:
      k8s-app: more-fs-watchers
  template:
    metadata:
      labels:
        name: more-fs-watchers
        k8s-app: more-fs-watchers
    spec:
      initContainers:
        - name: sysctl
          image: alpine:3
          command:
            - sysctl
            - -w
            - fs.inotify.max_user_watches=16384
          resources:
            requests:
              cpu: 10m
              memory: 1Mi
            limits:
              cpu: 100m
              memory: 5Mi
          securityContext:
            runAsUser: 0
            privileged: true
            readOnlyRootFilesystem: true
      containers:
        - name: pause
          image: gcr.io/google_containers/pause-amd64:3.1
          command:
            - /pause
          resources:
            requests:
              cpu: 10m
              memory: 1Mi
            limits:
              cpu: 100m
              memory: 5Mi
          securityContext:
            allowPrivilegeEscalation: false
            runAsUser: 1000
            runAsNonRoot: true
            readOnlyRootFilesystem: true
```